### PR TITLE
Optimize tagging for a large number of data sources (100K+)

### PR DESCRIPTION
**Pull request by timt51**
Optimize tagging process when only a small percentage of tags have changed.

Currently, the tagging script will make an Immuta API call at least once for each data source, and will make an API call to update a data dictionary (defining the column tags) regardless of whether or not the data dictionary has changed. When a data dictionary is updated via API call, regardless of whether or not the data dictionary has changed, Immuta will run Snowflake commands to update the data source's policies. Both of these processes take an excessively long time (>>4 hours) when the number of data sources is on the order of ~200K.

This PR optimizes the tagging script by updating data source tags and data dictionaries only when the set of tags has changed. When no tags have changed, the previous script takes ~6.5 minutes / 2.4K data sources, and the new script takes ~1.5 minutes / 2.4 K data sources. We expect an even larger difference as the number of data sources increase.

This PR also updates the tagging script so that it removes extra data source tags (it would only add new tags before).

Testing Strategy: Manual - test that the script appropriately adds and removes both data source and column tags.

### CODE REVIEW

**Review by briantist — COMMENTED** [review]
I don't have much familiarity with this code base so I'll leave it to Zak and Chintan to review more deeply. I would like to see more test coverage though.

I'll only add a recommendation to use the latest versions of the actions in the CI.

`actions/checkout`, `actions/setup-python`, and `codecov/codecov-action` are all up to `v3`.

That could be done in a different PR too, just mentioning.

File: fh_immuta_utils/tagging.py [hunk 1]
```diff
@@ -19,6 +17,13 @@
 LOGGER = logging.getLogger(__name__)
 
 
+# Tags created by Immuta by default
+IMMUTA_SPECIAL_TAGS = {
```
Comment (fh-katherinebargar) [inline fh_immuta_utils/tagging.py]: Do we know how often these are changed by Snowflake?

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -1,10 +1,10 @@
...
 - repo: https://github.com/pre-commit/pre-commit-hooks
-  rev: master
+  rev: main
   hooks:
     - id: trailing-whitespace
 - repo: https://github.com/python/black
-  rev: 21.6b0
+  rev: 22.3.0
   hooks:
     - id: black
```
Comment (fh-katherinebargar) [inline .pre-commit-config.yaml:9]: Maybe out of scope, but we usually set line-length to 120.

File: fh_immuta_utils/scripts/tag_existing_data_sources.py [hunk 3]
```diff
@@ -5,15 +5,20 @@
...
 from tqdm import tqdm
 
 from fh_immuta_utils.client import get_client
 from fh_immuta_utils.config import parse_config
 from fh_immuta_utils.paginator import Paginator
-from fh_immuta_utils.tagging import Tagger
+from fh_immuta_utils.tagging import IMMUTA_SPECIAL_TAGS, Tagger
+
+
+IMMUTA_API_PAGE_SIZE = 25_000
```
Comment (fh-katherinebargar) [inline fh_immuta_utils/scripts/tag_existing_data_sources.py:21]: Nit: Maybe this constant belongs in the client.py file?

File: fh_immuta_utils/scripts/tag_existing_data_sources.py [hunk 4]
```diff
@@ -77,38 +83,156 @@ def main(
...
+    ):
+        tag_names = set(tagger.get_tags_for_column(column))
+        with Paginator(
+            client.get_data_source_list,
+            search_text=search_text,
+            search_schema=search_schema,
+            columns=[column],
+            size=IMMUTA_API_PAGE_SIZE,
+        ) as paginator:
+            for data_source in paginator:
```
Comment (fh-katherinebargar) [inline fh_immuta_utils/scripts/tag_existing_data_sources.py:189]: Could we do this work in the existing iteration over client.get_data_source_list? (line 71ish)

File: fh_immuta_utils/scripts/tag_existing_data_sources.py [hunk 5]
```diff
@@ -77,38 +83,156 @@ def main(
...
-            desc=f"Tagging ID: {data_source['id']}, Name: {data_source['name']} :"
-        )
+    logging.info("Getting current data source and column tags")
+    # data_source_name: {tag_1_name, tag_2_name, ...}
+    current_data_source_tag_names_map: Dict[str, Set[str]] = defaultdict(set)
+    # data_source_name: {column_1_name: {tag_1_name, tag_2_name, ...}, ...}
+    current_column_tag_names_map: Dict[str, Dict[str, Set[str]]] = defaultdict(
+        lambda: defaultdict(set)
+    )
+    for tag_info in client.get_all_data_source_and_column_tags()["hits"]:
```
Comment (fh-katherinebargar) [inline fh_immuta_utils/scripts/tag_existing_data_sources.py]: Does this endpoint support pagination or querying based on tag_values?

File: fh_immuta_utils/scripts/tag_existing_data_sources.py [hunk 6]
```diff
@@ -77,38 +83,156 @@ def main(
...
+        progress_iterator.set_description(
+            desc=f"[Tagging Data Sources] Tag: {tag_name}, # Data Sources: {len(data_source_ids)} :"
+        )
+        if not dry_run:
+            client.update_data_source_tags_in_bulk(
+                ids=list(data_source_ids),
+                tag_data=[
+                    {
+                        "name": tag_name,
+                        "source": "curated",
```
Comment (fh-katherinebargar) [inline fh_immuta_utils/scripts/tag_existing_data_sources.py]: What does "curated" mean?

File: fh_immuta_utils/client.py [hunk 7]
```diff
@@ -535,6 +548,20 @@ def tag_data_source(self, id: int, tag_data: List[Dict[str, Any]]) -> bool:
...
+        self, ids: List[int], tag_data: List[Dict[str, Any]]
+    ) -> bool:
+        """
+        Adds tags to a list of data sources.
+        :param ids: data source ids
+        :param tag_data: list of tag dicts to apply to the data sources
+        :return: True
+        """
+        self.put(
+            "dataSource/bulk/tags", data={"ids": ids, "update": {"tags": tag_data}}
```
Comment (nkrishnan) [inline fh_immuta_utils/client.py:582]: i.e. 1:1 correspondence between list of ids and list of tag_data?

Comment (nkrishnan) [reply fh_immuta_utils/client.py:582]: nvm, from the usage it appears apply all tags to all ids in the list

File: fh_immuta_utils/scripts/tag_existing_data_sources.py [hunk 8]
```diff
@@ -77,38 +83,156 @@ def main(
...
+                wanted_column_tag_names_map[data_source_name][column] = tag_names
+
+    # Compute the data sources with column tags to update by checking if the current
+    # data dictionary matches the data dictionary we want.
+    data_sources_with_column_tags_to_update = set()
+    for data_source in data_sources_to_tag:
+        data_source_name = data_source["name"]
+        data_source_id = data_source["id"]
+        if (
+            current_column_tag_names_map[data_source_name]
```
Comment (fh-katherinebargar) [inline fh_immuta_utils/scripts/tag_existing_data_sources.py:200]: This is going to be confusing, lmk if you want want to zoom.

We're refreshing all the column's tags on the data source if one of the column's tags is out of sync. I feel like there's some kind of clever shortcut we could use to avoid loading and comparing all the tags. Something like hashing columns + tags, or lazily checking a data_source's columns and stopping once we know it has at least one mismatch.

**Comment by fh-katherinebargar** [comment]
I think I agree with the codecov bot. There's a lot of new code here that isn't tested.

One strategy would be to pull the behavior out of the script into the tagger (or some other new module). Then you could easily test the various edge cases of "check for differences and update in these scenarios" by mocking out the calls to Immuta. The script would still be a little messy but it would be calling into a well tested core.

File: fh_immuta_utils/tagging.py (hunk 1 above)
Comment (timt51) [reply fh_immuta_utils/tagging.py]: these are created by Immuta. It should change very infrequently if ever.

File: fh_immuta_utils/scripts/tag_existing_data_sources.py (hunk 3 above)
Comment (timt51) [reply fh_immuta_utils/scripts/tag_existing_data_sources.py:21]: the idea was that this value probably works well for this script, but maybe not for other use cases. but this seems like a decent default value to me so maybe making it the default would be better 🤷

File: fh_immuta_utils/scripts/tag_existing_data_sources.py (hunk 6 above)
Comment (timt51) [reply fh_immuta_utils/scripts/tag_existing_data_sources.py]: not really sure and it's not terribly important... at the very least any tag of interest to us is "curated". Switched to using the `Tag` object which sets the source to curated by default

File: fh_immuta_utils/scripts/tag_existing_data_sources.py (hunk 4 above)
Comment (timt51) [reply fh_immuta_utils/scripts/tag_existing_data_sources.py:189]: I don't think so, b/c you'd miss data sources with no matching column (but should still have a data source level tag)

File: fh_immuta_utils/scripts/tag_existing_data_sources.py (hunk 8 above)
Comment (timt51) [reply fh_immuta_utils/scripts/tag_existing_data_sources.py:200]: Sounds like it has potential... but I'm not sure it's worth optimizing further. Happy to discuss it more tho if you'd like

File: fh_immuta_utils/client.py (hunk 7 above)
Comment (timt51) [reply fh_immuta_utils/client.py:582]: yep - added to the docstring to clarify this.

**Comment by timt51** [comment]
> I think I agree with the codecov bot. There's a lot of new code here that isn't tested.
> 
> One strategy would be to pull the behavior out of the script into the tagger (or some other new module). Then you could easily test the various edge cases of "check for differences and update in these scenarios" by mocking out the calls to Immuta. The script would still be a little messy but it would be calling into a well tested core.

agreed! i dont think im going to get to this to though so I may leave it to the team to pick up on it

File: .pre-commit-config.yaml (hunk 2 above)
Comment (timt51) [reply .pre-commit-config.yaml:9]: i think the consensus on the team a long while back was to use black's defaults - but not sure

**Review by nkrishnan — APPROVED** [review]
